horovod/common/operations.cc

```diff
@@ -2135,94 +2136,100 @@ bool RunLoopOnce(HorovodGlobalState& state, bool is_coordinator) {
-          auto& entry = state.tensor_table[response.tensor_names()[0]];
-          tensor_size = entry.tensor->size();
-
-          while (!responses.empty()) {
-            auto new_response = responses.front();
-            assert(new_response.tensor_names().size() == 1);
-            auto& new_entry =
-                state.tensor_table[new_response.tensor_names()[0]];
-            int64_t new_tensor_size = new_entry.tensor->size();
-
-            if (response.response_type() == new_response.response_type() &&
-                response.devices() == new_response.devices() &&
-                entry.tensor->dtype() == new_entry.tensor->dtype() &&
-                tensor_size + new_tensor_size <= TensorFusionThresholdBytes()) {
-              // These tensors will fuse together well.
-              tensor_size += new_tensor_size;
-              response.add_tensor_name(new_response.tensor_names()[0]);
-              responses.pop_front();
-            } else {
-              // Don't try to fuse additional tensors since they are usually
-              // computed in order of requests and skipping tensors may mean
-              // that the batch will have to wait longer while skipped tensors
-              // could be reduced at that time.
-              break;
+      for (auto& r: responses_by_dtype) {
+        auto& responses = r.second;
+        while (!responses.empty()) {
+
+          auto response = responses.front();
+          assert(response.tensor_names().size() == 1);
+          responses.pop_front();
+          int64_t tensor_size = 0;
+
+          if (response.response_type() == MPIResponse::ResponseType::ALLREDUCE) {
+            // Attempt to add more responses to this fused response.
+            auto& entry = state.tensor_table[response.tensor_names()[0]];
+            tensor_size = entry.tensor->size();
+
+            while (!responses.empty()) {
+              auto new_response = responses.front();
+              assert(new_response.tensor_names().size() == 1);
+              auto& new_entry =
+                  state.tensor_table[new_response.tensor_names()[0]];
+              int64_t new_tensor_size = new_entry.tensor->size();
+
+              if (response.response_type() == new_response.response_type() &&
+                  response.devices() == new_response.devices() &&
+                  entry.tensor->dtype() == new_entry.tensor->dtype() &&
+                  tensor_size + new_tensor_size <= TensorFusionThresholdBytes()) {
+                tensor_size += new_tensor_size;
+                response.add_tensor_name(new_response.tensor_names()[0]);
+                responses.pop_front();
+              } else {
+                // Don't try to fuse additional tensors since they are usually
+                // computed in order of requests and skipping tensors may mean
+                // that the batch will have to wait longer while skipped tensors
+                // could be reduced at that time.
+                break;
```

Comment:
I did know this would lead to some misordering, as the map data structure will always force all FP16 ops to take precedence over FP32 operations, which in general might not be a nice thing to do. Your suggested version would definitely fix that. As a side effect, this change would introduce look ahead across response type boundaries though (I don't think this is necessarily a bad thing, but if you think the break condition should also maintain this behavior, let me know).

Reply:
I think we could introduce look-ahead penalty expressed in terms of size, and approximate the cost of obtaining the output by its size.  Then, we could break once the size of all skipped tensors is greater than X.  I don't know the good value for X, maybe 50% of fusion buffer size?  Or 10%... or 100%...  Worth exploring, and perhaps expose another tuning parameter.

Reply:
Or possibly, we could continue while `sum(existing output size) + sum(skipped size) <= TensorFusionThresholdBytes()`.  That would avoid another hyperparameter and still feels good.

Reply:
I agree on limiting based on sum of skipped messages (much better than adding another tunable parameter too). I went ahead a pushed a change to do exactly this so please take a look.

It should have mostly no impact on my current benchmark case since this logic should result in the same fusing (total gradient size per step is under the default 64 MiB fusion threshold so each cycle should launch two fully fused allreduces, one FP16 and one FP32). The order that these two operations occur might change, as it now will honor whatever response was first in queue. 

The more preserved ordering may improve the single node cases with modified `HOROVOD_FUSION_THRESHOLD` so I will rerun those tests to see.